chain-impl-mockchain/benches/tally.rs

```diff
@@ -18,101 +18,136 @@ use chain_impl_mockchain::{
-        wallets.push(wallet_builder);
-    }
+    wallets.append(&mut voters_wallets.iter_mut().collect());
 
-    let mut rng = TestCryptoRng::from_seed([0u8; 16]);
+    // Prepare committee members keys
     let members = CommitteeMembersManager::new(&mut rng, THRESHOLD, MEMBERS_NO);
-
-    let committee_keys = members
+    let committee_keys: Vec<_> = members
         .members()
         .iter()
         .map(|committee_member| committee_member.public_key())
-        .collect::<Vec<_>>();
+        .collect();
+
+    // Build the vote plan
+    let mut vote_plan_builder = vote_plan(VOTE_PLAN);
+    vote_plan_builder
+        .owner(ALICE)
+        .consecutive_epoch_dates()
+        .payload_type(PayloadType::Private)
+        .committee_keys(committee_keys);
+    for _ in 0..n_proposals {
+        let mut proposal_builder = proposal(VoteTestGen::external_proposal_id());
+        proposal_builder.options(3).action_parameters_no_op();
+        vote_plan_builder.with_proposal(&mut proposal_builder);
+    }
 
+    // Initialize ledger
     let (mut ledger, controller) = prepare_scenario()
         .with_config(
             ConfigBuilder::new(0)
                 .with_fee(LinearFee::new(0, 0, 0))
                 .with_rewards(Value(1000)),
         )
         .with_initials(wallets)
-        .with_vote_plans(vec![vote_plan(VOTE_PLAN)
-            .owner(ALICE)
-            .consecutive_epoch_dates()
-            .payload_type(PayloadType::Private)
-            .committee_keys(committee_keys)
-            .with_proposal(
-                proposal(VoteTestGen::external_proposal_id())
-                    .options(3)
-                    .action_transfer_to_rewards(100),
-            )])
+        .with_vote_plans(vec![&mut vote_plan_builder])
         .build()
         .unwrap();
 
-    let mut alice = controller.wallet(ALICE).unwrap();
+    // cast votes
+    let favorable = Choice::new(1);
 
     let vote_plan_def = controller.vote_plan(VOTE_PLAN).unwrap();
     let vote_plan: VotePlan = vote_plan_def.clone().into();
-    let proposal = vote_plan_def.proposal(0);
-
-    for alias in voters_aliases {
-        let mut private_voter = controller.wallet(&alias).unwrap();
-
-        controller
-            .cast_vote_private(
-                &private_voter,
-                &vote_plan_def,
-                &proposal.id(),
-                favorable,
-                &mut ledger,
-                &mut rng,
-            )
-            .unwrap();
-        private_voter.confirm_transaction();
+
+    let should_vote_distribution =
+        WeightedIndex::new(&[1.0 - proposals_per_voter_ratio, proposals_per_voter_ratio]).unwrap();
+
+    let mut total_votes_per_proposal = vec![0; n_proposals];
+
+    for (i, proposal) in vote_plan.proposals().iter().enumerate() {
+        for (alias, voting_power) in voters_aliases.iter().zip(voting_powers.iter()) {
+            let should_vote = should_vote_distribution.sample(&mut rng);
+            if should_vote == 0 {
```

Comment:
I think this is simpler https://docs.rs/rand/0.8.3/rand/trait.Rng.html#method.gen_bool

Reply:
`WeightedIndex` doesn't implement `Rng` and we need a particular distribution here. So we cannot do that unfortunately.

Reply:
But we already have a `Rng`  https://github.com/input-output-hk/chain-libs/blob/f5791cd3597709fe981851e7b22ca00ec9168cf3/chain-impl-mockchain/benches/tally.rs#L41. As far as I understand weighted index is only used to get a boolean that is true with certain probability, that is exactly what `rng::gen_bool` does

Reply:
Ouch, I missed the part about the probability parameter in `gen_bool`. You're right!